### PR TITLE
Improved Camera Movement

### PR DESCRIPTION
**Pull request by PetTurtle**
Created CameraControls node with all camera movement logic
Added camera drag with middle mouse
Added camera zoom to mouse position
Refactored movement

### CODE REVIEW

File: Assets/Player/CameraControls.gd [hunk 1]
```diff
@@ -0,0 +1,114 @@
...
+const MOVE_SPEED = 10
+const MOVE_FASTER_MULT = 3
+
+export(NodePath) var origin_path : NodePath
+export(NodePath) var camera_path : NodePath
+export(NodePath) var rotation_y_path : NodePath
+
+onready var _origin : Spatial = get_node(origin_path)
+onready var _camera : Camera = get_node(camera_path)
+onready var _rotation_y : Spatial = get_node(rotation_y_path)
```
Comment (aaronfranke) [inline Assets/Player/CameraControls.gd]: When explicitly stating the type, do not include a space before the colon. (but do include a preceding space for `:=`)

File: Assets/Player/CameraControls.gd [hunk 2]
```diff
@@ -0,0 +1,114 @@
...
+
+
+func _move_drag():
+	if Input.is_action_pressed("move_drag"):
+		var m_pos = _viewport.get_mouse_position()
+		if Input.is_action_just_pressed("move_drag"):
+			_move_drag_start = m_pos
+		else:
+			var drag_dir = _move_drag_start - m_pos
+			drag_dir /= _viewport_size / _camera.size
```
Comment (aaronfranke) [inline Assets/Player/CameraControls.gd]: Division is an expensive operation and should be minimized. Do this instead:

    drag_dir *= _camera.size / _viewport_size

File: Assets/Player/CameraControls.gd [hunk 3]
```diff
@@ -0,0 +1,114 @@
...
+	if _viewport.connect("size_changed", self, "_on_viewport_size_changed") != OK:
+		push_error("Failed To Connect Viewport")
+	_basis = _get_basis()
+
+func _process(delta):
+	_move(delta)
+	_move_drag()
+
+
+func _input(event):
```
Comment (aaronfranke) [inline Assets/Player/CameraControls.gd]: While the GDScript style guide does state to use 2 lines between functions, Unknown Horizons uses 1 line.

File: Assets/Player/CameraControls.gd [hunk 4]
```diff
@@ -0,0 +1,114 @@
...
+
+
+func set_enabled(new_value):
+	set_process(new_value)
+	set_process_input(new_value)
+	enabled = new_value
+
+
+func get_enabled() -> bool:
+	return enabled;
```
Comment (aaronfranke) [inline Assets/Player/CameraControls.gd]: Unnecessary semicolon.

**Review by aaronfranke — CHANGES_REQUESTED** [review]

File: Assets/Player/CameraControls.tscn [hunk 5]
```diff
@@ -0,0 +1,3 @@
+[gd_scene format=2]
+
+[node name="CameraControls" type="Node"]
```
Comment (aaronfranke) [inline Assets/Player/CameraControls.tscn]: This should not be a scene.

**Review by aaronfranke — CHANGES_REQUESTED** [review]

**Review by aaronfranke — APPROVED** [review]
Style looks good. @artism90 Please review.

File: Assets/Player/CameraControls.gd [hunk 6]
```diff
@@ -0,0 +1,103 @@
...
+onready var _rotation_y := get_node(rotation_y_path) as Spatial
+onready var _viewport := get_viewport() as Viewport
+onready var _viewport_size := _viewport.size as Vector2
+onready var _viewport_aspect := _viewport_size.aspect() as float
+
+var _basis: Basis
+var _move_drag_start: Vector2
+var enabled: bool = true setget set_enabled, get_enabled
+
+func _ready():
```
Comment (pingwindev) [inline Assets/Player/CameraControls.gd]: Add parameter types and return types to all functions. If there have been existing functions without them, it was most likely an oversight.

`func _ready() -> void:`

File: Assets/Player/CameraControls.gd [hunk 7]
```diff
@@ -0,0 +1,103 @@
...
+	if _viewport.connect("size_changed", self, "_on_viewport_size_changed") != OK:
+		push_error("Failed To Connect Viewport")
+	_basis = _get_basis()
+
+func _process(delta):
```
Comment (pingwindev) [inline Assets/Player/CameraControls.gd]: `func _process(delta: float) -> void:`

File: Assets/Player/CameraControls.gd [hunk 8]
```diff
@@ -0,0 +1,103 @@
...
+	_move(delta)
+	_move_drag()
+
+func _input(event):
```
Comment (pingwindev) [inline Assets/Player/CameraControls.gd]: `func _input(event: InputEvent) -> void:`

File: Assets/Player/CameraControls.gd [hunk 9]
```diff
@@ -0,0 +1,103 @@
...
+	
+	elif event.is_action_pressed("zoom_in"):
+		if _camera.size > ZOOM_IN_LIMIT:
+			_zoom(-ZOOM_VALUE)
+	
+	elif event.is_action_pressed("zoom_out"):
+		if _camera.size < ZOOM_OUT_LIMIT:
+			_zoom(ZOOM_VALUE)
+
+func _move(delta: float):
```
Comment (pingwindev) [inline Assets/Player/CameraControls.gd]: `func _move(delta: float) -> void:`

File: Assets/Player/CameraControls.gd [hunk 10]
```diff
@@ -0,0 +1,103 @@
...
+	var movement_scale: float = delta * MOVE_SPEED
+	if Input.is_action_pressed("move_faster"):
+		movement_scale *= MOVE_FASTER_MULT
+	
+	var x = Input.get_action_strength("move_right") - Input.get_action_strength("move_left")
+	var y = Input.get_action_strength("move_down") - Input.get_action_strength("move_up")
+	var movement_velocity = Vector3(x, 0, y * _viewport_aspect)
+	_origin.translate(_basis.xform(movement_velocity) * movement_scale)
+
+func _move_drag():
```
Comment (pingwindev) [inline Assets/Player/CameraControls.gd]: `func _move_drag() -> void:`

File: Assets/Player/CameraControls.gd [hunk 11]
```diff
@@ -0,0 +1,103 @@
...
+		var m_pos = _viewport.get_mouse_position()
+		if Input.is_action_just_pressed("move_drag"):
+			_move_drag_start = m_pos
+		else:
+			var drag_dir = (_move_drag_start - m_pos) * _camera.size / _viewport_size
+			var move_dir = _basis.xform(Vector3(drag_dir.x, 0, drag_dir.y))
+			_origin.translate(move_dir)
+			_move_drag_start = m_pos
+
+func _rotate(rotation: float):
```
Comment (pingwindev) [inline Assets/Player/CameraControls.gd]: `func _rotate(rotation: float) -> void:`

File: Assets/Player/CameraControls.gd [hunk 12]
```diff
@@ -0,0 +1,103 @@
...
+	_rotation_y.rotate_y(rotation)
+	_basis = _get_basis()
+
+func _zoom(zoom_value: float):
```
Comment (pingwindev) [inline Assets/Player/CameraControls.gd]: `func _zoom(zoom_value: float) -> void:`

File: Assets/Player/CameraControls.gd [hunk 13]
```diff
@@ -0,0 +1,103 @@
...
+func _raycast_from_mouse(m_pos: Vector2, collision_mask: int) -> Dictionary:
+	var ray_start = _camera.project_ray_origin(m_pos)
+	var ray_end = ray_start + _camera.project_ray_normal(m_pos) * RAY_LENGTH
+	var space_state = _origin.get_world().direct_space_state
+	return space_state.intersect_ray(ray_start, ray_end, [], collision_mask)
+
+func _get_basis() -> Basis:
+	return _rotation_y.get_transform().basis
+
+func set_enabled(new_value: bool):
```
Comment (pingwindev) [inline Assets/Player/CameraControls.gd]: `func set_enabled(new_value: bool) -> void:`

File: Assets/Player/CameraControls.gd [hunk 14]
```diff
@@ -0,0 +1,103 @@
...
+	set_process(new_value)
+	set_process_input(new_value)
+	enabled = new_value
+
+func get_enabled() -> bool:
+	return enabled
+
+func _on_viewport_size_changed():
```
Comment (pingwindev) [inline Assets/Player/CameraControls.gd]: `func _on_viewport_size_changed() -> void:`

**Review by pingwindev — CHANGES_REQUESTED** [review]
Drag speed needs to be tweaked for higher resolutions. It's usable in a 1600x900 window but a bit slow on 1920p and way too slow on 4K. This can be addressed in a future PR though.

Please fix the mentioned code issues and rebase to the latest state.

**Pull request closed without merging**